src/foam/dao/QueryCachingDAODecorator.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'QueryCachingDAODecorator',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: 'Javascript DAO Decorator which adds select caching to a delegate DAO.',
+
+  constants: [
+    {
+      name: 'CLEAR_CACHE',
+      value: 'CLEAR_CACHE'
+    }
+  ],
+
+  properties: [
+    {
+      // The cache for local storage and fast access
+      name: 'cache',
+      factory: function() { return {}; }
+    },
+    {
+      // Number of elements in the DAO backing this query cache
+      // Allows limit to what we cache so we do not have to always call DAO if limit is not provided
+      name: 'daoCount_',
```

Comment:
Why do we need daoCount_?

Reply:
Without knowing how many elements are in the DAO list, the cache would always have to call the DAO to make sure there were no entries after the last entry in the cache (e.g. when limit is not passed in)

Reply:
Why do we care? This is intended to work for skip() and limit() when scrolling. If someone is doing a full select(), then they must expect the DAO to be small enough for that to work, and if it doesn't, then that's their problem. If they didn't use skip and limit, then just looking for a few extra rows isn't safe because based on the sort order, the new record may not be at the end, but somewhere in the middle and asking beyond the previous limit will probably just give you values you already had. This cache is just useful for short-term non-critical caching when skip() and limit() are used. If skip and limit aren't specified, it can just delegate.

Reply:
I have removed daoCount_ and delegate select call if limit not provided